prime-router/src/main/kotlin/tokens/TokenAuthentication.kt

```diff
@@ -0,0 +1,220 @@
+                    logger.error("AccessToken check failed - no claims.  Unauthorized.")
+                    return null
+                }
+                if (isExpiredToken(jws.body.expiration)) {
+                    logger.error("AccessToken expired.  Unauthorized.")
+                    return null
+                }
+                val scope = jws.body["scope"] as? String
+                if (scope == null) {
+                    logger.error("Missing scope claim.  Unauthorized.")
+                    return null
+                }
+                if (!Scope.scopeListContainsScope(scope, desiredScope)) {
+                    logger.error("Signed key has scope $scope, but requests $desiredScope.  Unauthorized")
+                    return null
+                }
+                return jws.body
+            } catch (ex: JwtException) {
+                logger.error("Unauthorized: $ex")
+                return null
+            } catch (exc: RuntimeException) {
+                logger.error("Unauthorized: $exc")
+                return null
+            }
+        }
+
+    companion object: Logging {
+        fun isExpiredToken(exp: Date): Boolean {
+            return (Date().after(exp))   // no need to include clock skew, since we generated token ourselves
+        }
+    }
+}
+
+/**
+ * Defined per the FHIR standard
+ *    https://hl7.org/fhir/uv/bulkdata/authorization/index.html
+ */
+// todo get this to work:  @JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
+data class AccessToken(
+    @JsonProperty("access_token")
+    val accessToken: String,   // access_token - required - The access token issued by the authorization server.
+    @JsonProperty("token_type")
+    val tokenType: String,     // token_type	required - Fixed value: bearer.
+    @JsonProperty("expires_in")
+    val expiresIn: Int,        // seconds until expiration, eg, 300
+    @JsonProperty("expires_at_seconds")
+    val expiresAtSeconds: Int, //  unix time IN SECONDS, of the expiration time.
+    @JsonProperty("scope")
+    val scope: String, // scope	required - Scope of access authorized. can be different from the scopes requested
+)
+
+/**
+ * This is used during validation of a SenderToken.
+ *
+ * Implementation of a callback function used to find the public key for
+ * a given Sender, kid, and alg.   Lookup in the Settings table.
+ *  todo:  the FHIR spec calls for allowing a set of keys. However, this callback only allows for one.
+ */
+class FindSenderKeyInSettings(val scope: String) : SigningKeyResolverAdapter(), Logging {
+    var errorMsg: String? = null
+
+    fun fail(shortMsg: String): Key? {
+        errorMsg = "Token Request Denied: Error while requesting $scope: $shortMsg"
+        logger.error(errorMsg!!)
+        return null
+    }
+
+    override fun resolveSigningKey(jwsHeader: JwsHeader<*>?, claims: Claims): Key? {
+        errorMsg = null
+        if (jwsHeader == null) return fail("JWT has missing header")
```

Comment:
A style point.... `fail` is part of the Kotlin.test library, meaning it should only be used for test code. Also, `return` is not necessary because `fail` always throws an exception.  

Suggest as more clear. 
```
checkNotNull(jwsHeader, "JWT has a missing header")
```

Reply:
Fixed.